test/SILOptimizer/noimplicitcopy_borrowing_parameters.swift

```diff
@@ -0,0 +1,574 @@
+    let _ = x // expected-note {{consumed here}}
+    let _ = x.t
+}
+
+func testAddressOnlyBorrowingConsume2<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    // expected-error @-2 {{'x' is borrowed and cannot be consumed}}
+    var y = x // expected-note {{consumed here}}
+    y = x // expected-note {{consumed here}}
+    _ = y
+}
+
+func testAddressOnlyBorrowingConsume3<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    let y = x // expected-note {{consumed here}}
+    _ = y
+}
+
+func testAddressOnlyBorrowingUse<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    _ = x
+}
+
+func testAddressOnlyBorrowingUseAndConsume<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    borrowValDefault(x)
+    let _ = x // expected-note {{consumed here}}
+}
+
+func testAddressOnlyBorrowingCallBorrowValDefault<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    borrowValDefault(x)
+}
+
+func testAddressOnlyBorrowingCallBorrowValBorrowing<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    borrowValBorrowing(x)
+}
+
+func testAddressOnlyBorrowingCallMethodSelfDefault<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    x.doSomethingDefault()
+}
+
+func testAddressOnlyBorrowingCallMethodSelfBorrowing<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    x.doSomethingBorrowing()
+}
+
+func testAddressOnlyBorrowingEscapingClosure<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    var f: () -> () = {}
+    f = { // expected-note {{consumed here}}
+        _ = x
+    }
+    _ = f
+}
+
+func testAddressOnlyBorrowingNonEscapingClosure<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    func useNonEscaping(_ f: () -> ()) {}
+    useNonEscaping { // expected-note {{consumed here}}
+        _ = x
+    }
+}
+
+func testAddressOnlyBorrowingCast<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    let _ = x as Any // expected-note {{consumed here}}
+}
+
+func testAddressOnlyBorrowingCastCheck<T : P>(_ x: borrowing GenericNonTrivialStruct<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    if x is Any { // expected-note {{consumed here}}
+        // expected-warning @-1 {{'is' test is always true}}
+
+    }
+}
+
+func testAddressOnlyBorrowingEnum<T>(_ x: borrowing AddressOnlyEnum<T>) {
+    // expected-error @-1 {{'x' is borrowed and cannot be consumed}}
+    switch x { // expected-note {{consumed here}}
+    case let .x(y):
+        _ = y
+        break
+    case let .y(z):
+        _ = z
+        break
+    }
+}
```

Comment:
I think we really need to tune these diagnostics to tell users that they can and are required to make an explicit copy.